server/src/main/java/org/opensearch/cluster/coordination/Coordinator.java

```diff
@@ -1354,6 +1362,9 @@ assert getLocalNode().equals(clusterState.getNodes().get(getLocalNode().getId())
                 currentPublication = Optional.of(publication);
 
                 final DiscoveryNodes publishNodes = publishRequest.getAcceptedState().nodes();
+                // marking pending disconnects before publish
+                // if a nodes tries to send a joinRequest while it is pending disconnect, it should fail
+                transportService.setPendingDisconnections(new HashSet<>(clusterChangedEvent.nodesDelta().removedNodes()));
```

Comment:
Instead of calling transportService directly for tracking pending disconnection. I have following suggestion

In ```ClusterConnectionManager.java```  via ```ConnectionManager``` interface add
```
    @Override
    public void trackPendingDisconnection(DiscoveryNode node) {
        //
    }
```
 and add same method in TransportService.java as well.
Then in NodeConnectionsService.java update the disconnect() method to
```
     Runnable disconnect() {
            transportService.trackPendingDisconnection(discoveryNode);   // track pending disconnection
            return addListenerAndStartActivity(
                null,
                ActivityType.DISCONNECTING,
                disconnectActivity,
                "connection cancelled by disconnection"
            );
        }
```

Reply:
I don't think this would work

The idea currently is to mark the pending disconnect at the start of node-left task _before state publication_.
This blocks any node-join tasks from getting queued until node-left task has completed state publication and state commit.

This `disconnect()` runnable runs as part of state commit phase of node-left task which happens after state publication.
So if we mark the disconnect as pending as part of state commit, it could be too late and join request might have already succeeded and queued a node-join task.

Scenario:
1. some node leaves due to a legitimate reason and triggers a node-left task
2. node-left task starts - it updates followerchecker and leaderchecker and starts state publication. 
3. leadercheck requests from leaving node now get rejected by cluster-manager.
4. leaving node sees that leader checks got rejected and goes into candidate mode and starts discovery.
5. leaving node sends a join request when it finds the active cluster-manager.
6. join request succeeds and queues a node-join task into cluster-manager
7. now node-left state publication completes and begins state commit.
8. node-left task marks the leaving node as pending disconnection, wipes the connection, completes the disconnection and removes the pending disconnection entry. node-left task then completes state commit.
9. node-join task starts processing and fails state publication because the connection was wiped.

In this scenario, the join request succeeds before the connection is wiped by node-left state commit. This is why we get into the race condition.
Instead, we need to ensure that join requests fail at any point while node-left is ongoing to avoid the race condition.

In order to do this, we need to mark the pending disconnect before state publication itself.
This wouldn't be possible if we marked as part of disconnect runnable - it has to be done in `Coordinator`

Let me know your thoughts

Reply:
got your point. should we move this ```nodeConnectionsService.disconnectFromNodesExcept(newClusterState.nodes());``` call itself in publish method directly?

Reply:
This would probably work for cluster-manager since it handles publication, but then other nodes would not disconnect from the leaving node once the new state is committed so we might see some other issues.

I think this is why it is currently being done as part of state commit phase - since it runs on all nodes.